### ISSUE TITLE
Clicking wikipedia-link with a section does not always work 

### ISSUE THREAD

**Issue by daganzdaanda** [openstreetmap/openstreetmap-website#967]
Hi, I have noticed something when looking at the tags of an object: 
Wikipedia links that point to sections (with # ) don't always work.

http://www.openstreetmap.org/way/130469169
has the tag
wikipedia=de:Englischer Garten (München)#Japanisches Teehaus

It's a clickable link to
`http://de.wikipedia.org/wiki/de:Englischer%20Garten%20(München)?uselang=de#Japanisches Teehaus`
This brings up the page, but does not jump to the section.

The section is reached by this link instead:
<http://de.wikipedia.org/wiki/Englischer_Garten_(München)#Japanisches_Teehaus>
Interestingly, both JOSM and OverpassTurbo translate the tag into the working link. 

It seems that it has to do with the spaces and also non-ascii characters: 
Simple one-word-anchors without "Umlauts" seem to be fine: <http://www.openstreetmap.org/node/1594364342>, but with an Umlaut, it does not work: <http://www.openstreetmap.org/node/1383144922>

Maybe loosely related is #712, that seems to be the last issue about wikipedia-links.

**Comment by tomhughes** [comment]
Well those are two completely different links - you have removed the `uselang` parameter!

**Comment by tomhughes** [comment]
Also, the link on the web site works find for me.

**Comment by tomhughes** [comment]
So this is arguably a bug in mediawiki in that it apparently translates space to underscore in the URL but not in the anchor.

**Comment by tomhughes** [comment]
To summarise, and ignoring all the irrelevant confusion about umlauts, and the various accidental modifications of unrelated bits of the URL in the initial report, the problem is that we generate this link:

```
http://de.wikipedia.org/wiki/de:Englischer%20Garten%20(M%C3%BCnchen)?uselang=en-GB#Japanisches%20Teehaus
```

when we need to generate:

```
http://de.wikipedia.org/wiki/de:Englischer%20Garten%20(M%C3%BCnchen)?uselang=en-GB#Japanisches_Teehaus
```

with an underscore in the anchor component, because mediawiki requires spaces to be translated to underscore for the anchor name when multiple words are used in a section heading.

The question is, is this really a bug in the web site core? or is the data in the database wrong? It seems that `de:Englischer Garten (München)#Japanisches Teehaus` is not really the right reference and it should be `de:Englischer Garten (München)#Japanisches_Teehaus`.

**Comment by daganzdaanda** [comment]
> you have removed the  uselang  parameter!

I just copied the urls, seems like OverpassTurbo did remove that.

>  irrelevant confusion about umlauts

Maybe its a separate issue, but there is an issue nonetheless.
As shown, an Umlaut in a section is not "escaped" correctly, so the link does not work.

Node: http://www.openstreetmap.org/node/1383144922
Tag: `wikipedia=de:Alte Brücke (Heidelberg)#Brückenaffe`

The link that the website generates does not work:
http://de.wikipedia.org/wiki/de:Alte%20Brücke%20(Heidelberg)?uselang=de#Brückenaffe

The link that OverpassTurbo generates does work:
http://de.wikipedia.org/wiki/Alte%20Brücke%20(Heidelberg)%23Brückenaffe
Note that it is transformed into this URL: 
http://de.wikipedia.org/wiki/Alte_Brücke_(Heidelberg)#Br.C3.BCckenaffe
An umlaut in the section breaks the link, but is OK in the main part?!

Adding the `uselang` breaks the link again:
http://de.wikipedia.org/wiki/Alte%20Brücke%20(Heidelberg)?uselang=de%23Brückenaffe

even escaping the "ü" in the section does not work:
http://de.wikipedia.org/wiki/Alte%20Brücke%20(Heidelberg)?uselang=de%23Br.C3.BCckenaffe

...except when you use the # instead of %23:
http://de.wikipedia.org/wiki/Alte%20Brücke%20(Heidelberg)?uselang=de#Br.C3.BCckenaffe

To sum it up:
* The section-part needs any umlauts (and ß) escaped, but with a dot instead of %. I haven't checked other non-ascii letters.
* The dash needs to be a dash and not escaped if there is a uselang-part
* Spaces need to be turned into underscores in the section part and could also be underscores in the main part


> The question is, is this really a bug in the web site code? or is the data in the database wrong?

The wiki currently says to 
*use the human readable title of the article, for example 'en:St Paul's Cathedral' rather than the version with underscores and html encoded characters that is used in the URL ('St_Paul%27s_Cathedral')*
It does not mention the special case of links to sections.

**Comment by tomhughes** [comment]
Well that unicode escaping for the sections is not something from any kind of standard - it is some entirely made up mediawiki thing.

So it's perfectly fine for the base URL part to be the readable form because (a) mediawiki accepts spaces there anyway and (b) the rest of it is just standard URL escaping. So we don't have to do anything - the data in the database is valid as is whichever form of title you use.

That is no true for the fragment, as wikipedia doesn't handle spaces or standard URL escaping there we would have to do special custom handling in order for human readable forms to work there. While certainly could do that in the web site code it wouldn't help the thousands of other users of our data.

**Comment by daganzdaanda** [comment]
Thank you, Tom! That was quick!

I found out one more thing: Commas need to be turned into ".2C" ...

<http://www.openstreetmap.org/way/188958935>

The current link does not work:
<http://de.wikipedia.org/wiki/de:Liste%20der%20Baudenkmäler%20in%20Eichstätt?uselang=de#Br.C3.BCckenstra.C3.9Fe_1,_Ehemaliges_Bauernhaus>

but with .2C replacing comma, it's working:
<http://de.wikipedia.org/wiki/de:Liste%20der%20Baudenkmäler%20in%20Eichstätt?uselang=de#Br.C3.BCckenstra.C3.9Fe_1.2C_Ehemaliges_Bauernhaus>


Really weird system at Wikipedia / Mediawiki. After a lot of searching, I found this explanation:
<http://www.mediawiki.org/wiki/Manual:PAGENAMEE_encoding#Encodings_compared>
The {{anchorencode:...}} line in the table gives the substitutions that are made.
I'm guessing what we currently do with URI.encode works like the {{urlencode:...|WIKI}} line?
If so, the differences are in the following characters: `!$()*,/;@` (Nice cartoon swearing...)
I actually found a real life wikipedia link with a / inside the section anchor:
<http://www.openstreetmap.org/way/30906354>

**Comment by tomhughes** [comment]
I actually based what I did on the mediawiki code, but it seems that the `urlencode` function in PHP is stricter that ruby (it actually does form encoding - the `rawurlencode` function does actual URL escaping).

**Comment by daganzdaanda** [comment]
Is the commit already live? 
The comma and the slash example still aren't working for me, and I found a semicolon that is also not working. 
comma: <http://www.openstreetmap.org/way/188958935>
slash: <http://www.openstreetmap.org/way/30906354>
semicolon: <http://www.openstreetmap.org/way/136026088>

Can't find any anchors that use one of the other special characters !$()*@ -- Either there are none, or OverpassTurbo can't find them because it seems to unescape things in regex before doing the search.

**Comment by tomhughes** [comment]
Oh no sorry, I left the tests running and forgot to mark it live when they finished. I've done it now, so give if half an hour and it will be up.

**Comment by daganzdaanda** [comment]
Now everything is OK! Thank you again, very good work!

**Comment by fgnievinski** [comment]
hi, it seems that diacritics also need some attention, please.

I've tried to add a link to 
https://pt.wikipedia.org/wiki/Tristeza_(Porto_Alegre)#Pórtico

when the non-ASCII character "ó" is escaped, the URL correctly becomes:
https://pt.wikipedia.org/wiki/Tristeza_(Porto_Alegre)#P%C3%B3rtico

but OSM is erroneously replacing the percent signs for dots:
https://pt.wikipedia.org/wiki/Tristeza_(Porto_Alegre)#P.C3.B3rtico

which mediawiki won't recognize when you click on a feature.
